README.rst
**⚠️ Current status of this project**

**This package has been renamed as `django-weasypdf` since v0.1.1, to avoid a conflict in PyPI, and moved to https://github.com/morlandi/django-weasypdf**.

You are strongly suggested to switch to the new project https://github.com/morlandi/django-weasypdf, which is actively maintained and provides several improvements;

however, all releases published in the current repo will rest in place forever.


django-pdf
==========

A Django class-based view helper to generate PDF with WeasyPrint.

Requires: `WeasyPrint <https://github.com/Kozea/WeasyPrint>`_

Optional requirements:

- matplotlib (to render plots)

.. contents::

.. sectnum::

Installation
------------

Install the package by running:

.. code:: bash

    pip install django-weasypdf

or:

.. code:: bash

    pip install git+https://github.com/morlandi/django-pdf


You will probably build you own app in the project to provide derived Views
and custom templates; for example:

.. code:: bash

    python manage.py startapp reports

In your settings, add:

.. code:: python

    INSTALLED_APPS = [
        ...
        'reports',
        'pdf',
    ]

Note that `reports` is listed before `pdf` to make sure you can possibly
override any template.

In your urls, add:

.. code:: python

    urlpatterns = [
        ...
        path('reports/', include('reports.urls', namespace='reports')),
        ...

You might want to copy the default templates from 'pdf/templates/pdf' to 'reports/templates/reports'
for any required customization; see `Customizing the templates`_ below

A sample report
---------------

A test view has been provided to render a sample report for demonstration purposes.

In your main urls, include pdf/urls.py, where the required end-point have been
mapped to the PdfTestView; then, visit:

    http://127.0.0.1:8000/pdf/test/print/

You can copy the following to your own app to have an initial working view
to start working with:

file `reports/urls.py`:

.. code:: python

    from django.urls import path
    from . import views

    app_name = 'pdf'

    urlpatterns = [
        path('test/print/', views.ReportTestView.as_view(), {'for_download': False, 'lines': 200, }, name="test-print"),
        path('test/download/', views.ReportTestView.as_view(), {'for_download': True, 'lines': 200, }, name="test-download"),
    ]


file `reports/views.py`:

.. code:: python

    from pdf.views import PdfView


    class ReportView(PdfView):

        #my_custom_data = None
        header_template_name = 'pdf/header.html'
        footer_template_name = 'pdf/footer.html'
        styles_template_name = 'pdf/styles.css'

        def get_context_data(self, **kwargs):
            context = super(ReportView, self).get_context_data(**kwargs)
            #self.my_custom_data = context.pop('my_custom_data', None)
            # context.update({
            #     'footer_line_1': config.REPORT_FOOTER_LINE_1,
            #     'footer_line_2': config.REPORT_FOOTER_LINE_2,
            # })
            return context


    class ReportTestView(ReportView):
        body_template_name = 'pdf/pages/test.html'
        styles_template_name = 'pdf/pages/test.css'
        # header_template_name = None
        # footer_template_name = None
        title = "Report Test"

        def get_context_data(self, **kwargs):
            context = super().get_context_data(**kwargs)

            # Add a plot
            try:
                from .plot import build_plot_from_data
                plot_image = build_plot_from_data(data=None, as_base64=True)
                context.update({
                    'plot_image': plot_image,
                })
            except:
                pass

            # Add your stuff here ...
            context.update({
                ...
            })

            return context


or **replace `pdf/header.html` with `reports/header.html`**, etc ... when using
custom templates.

file `reports/pages/test.html`:

.. code:: html

    {% extends "pdf/base.html" %}

    {% block content %}

        <h1>Test PDF</h1>

        {% if plot_image %}
            <img class="plot" src="data:image/png;base64,{{plot_image}}">
        {% endif %}

        {% with lines=lines|default:100 %}
            {% for i in "x"|rjust:lines %}
                <div>line {{forloop.counter}} ...</div>
            {% endfor %}
        {% endwith %}

    {% endblock content %}


You can now download the PDF document at:

    http://127.0.0.1:8000/reports/test/download/

or open it with the browser at:

    http://127.0.0.1:8000/reports/test/print/

You can inspect the HTML used for PDF rendering by appending `?format=html` to the url:

    http://127.0.0.1:8000/reports/test/print/?format=html

.. image:: screenshots/pdf_sample.png


Building a PDF document from a background process
-------------------------------------------------

A `PdfView.render_as_pdf_to_stream(self, base_url, extra_context, output)` method is supplied for this purpose:

.. code:: python

    def render_as_pdf_to_stream(self, base_url, extra_context, output):
        """
        Build the PDF document and save in into "ouput" stream.

        Automatically called when the view is invoked via HTTP (unless self.format == 'html'),
        but you can also call it explicitly from a background task:

            view = PdfTestView()
            context = view.get_context_data()
            with open(filepath, 'wb') as f:
                view.render_as_pdf_to_stream('', context, f)
        """

A sample management command to build a PDF document outside the HTML request/response
cycle is available here:

`pdf/management/commands/build_test_pdf.py <./pdf/management/commands/build_test_pdf.py>`_


Providing "extra_context" parameters
------------------------------------

Supply context parameters either in the urlpattern, or invoking get_context_data():

from `urls.py`:

.. code:: python

    urlpatterns = [
        path('daily/print/', views.ReportDailyView.as_view(), {'exclude_inactives': False}, name="daily-print"),
    ]

from a background task:

.. code:: python

    from django.core.files.base import ContentFile

    # Create a View to work with
    from reports.views import ReportDailyView
    view = ReportDailyView()
    context = view.get_context_data(
        exclude_inactives=task.exclude_inactives,
    )

    # Create empty file as result
    filename = view.build_filename(extension="pdf")
    task.result.save(filename, ContentFile(''))

    # Open and write result
    filepath = task.result.path

    with open(filepath, 'wb') as f:
        view.render_as_pdf_to_stream('', context, f)


Customizing the templates
-------------------------

These sample files::

    pdf
    ├── static
    │   └── pdf
    │       └── images
    │           └── header_left.png
    └── templates
        └── pdf
            ├── base.html
            ├── base_nomargins.html
            ├── styles.css
            ├── footer.html
            ├── header.html
            └── pages
                ├── test.css
                └── test.html


can be copied into your app's local folder `reports/templates/reports`,
and used for any required customization:

.. code:: python

    class ReportView(PdfView):

        header_template_name = 'reports/header.html'
        footer_template_name = 'reports/footer.html'
        styles_template_name = 'reports/styles.css'

How to insert a page break
--------------------------

.. code:: html

    <p style="page-break-before: always" ></p>


Adding Weasyprint to your project
---------------------------------

Add `weasyprint` to your requirements::

    WeasyPrint==51

and optionally to your LOGGING setting::

    LOGGING = {
        ...
        'loggers': {
            ...
            'weasyprint': {
                'handlers': ['console'],
                'level': 'DEBUG',
                'propagate': True,
            },
        },
    }

Deployment:

1) Install Courier fonts for PDF rendering

::

    # You can verify the available fonts as follows:
    #    # fc-list
    - name: Install Courier font for PDF rendering
        become: true
        become_user: root
        copy:
            src: deployment/project/courier.ttf
            dest: /usr/share/fonts/truetype/courier/

The font file can be downloaded here:

`courier.ttf <resources/fonts/courier.ttf>`_

2) You might also need to install the following packages:

::

    #weasyprint_packages:
    - libffi-dev          # http://weasyprint.readthedocs.io/en/latest/install.html#linux
    - python-cffi         # http://weasyprint.readthedocs.io/en/latest/install.html#linux
    - python-dev          # http://weasyprint.readthedocs.io/en/latest/install.html#linux
    - python-pip          # http://weasyprint.readthedocs.io/en/latest/install.html#linux
    - python-lxml         # http://weasyprint.readthedocs.io/en/latest/install.html#linux
    - libcairo2           # http://weasyprint.readthedocs.io/en/latest/install.html#linux
    - libpango1.0-0       # http://weasyprint.readthedocs.io/en/latest/install.html#linux
    - libgdk-pixbuf2.0-0  # http://weasyprint.readthedocs.io/en/latest/install.html#linux
    - shared-mime-info    # http://weasyprint.readthedocs.io/en/latest/install.html#linux
    - libxml2-dev         # http://stackoverflow.com/questions/6504810/how-to-install-lxml-on-ubuntu#6504860
    - libxslt1-dev        # http://stackoverflow.com/questions/6504810/how-to-install-lxml-on-ubuntu#6504860

For an updated list, check here:

https://weasyprint.readthedocs.io/en/latest/install.html#linux

Customizations examples
=======================

Add a custom logo
-----------------

For example you can save a custom bitmap with django-constance:

.. code :: python

    CONSTANCE_ADDITIONAL_FIELDS = {
        'image_field': ['django.forms.ImageField', {}]
    }

    CONSTANCE_CONFIG = {
        ...
        'PDF_RECORD_LOGO': ('', 'Image for PDF logo', 'image_field'),
    }

then in your **header.html** template:

.. code:: html

    <body>
        <div class="pageHeader">
            <img class="pageLogo" title="{{ PDF_RECORD_LOGO }}" src="media://{{ PDF_RECORD_LOGO }}">
            <div class="pageTitle">{{print_date|date:'d/m/Y H:i:s'}} - {{title}}</div>
        </div>
    </body>

Embed images from media (ImageFields)
-------------------------------------

If Image is a Model to keep the images you want to embed, use a templatetag like this:

.. code:: python

    @register.filter
    def local_image_url(image_slug):
        """
        Example:
            "/backend/images/signature_mo.png"
        """

        url = ''
        try:
            image = Image.objects.get(slug=image_slug)
            if bool(image.image):
                url = image.image.url.lstrip(settings.MEDIA_URL)
        except Image.DoesNotExist as e:
            pass

        if len(url):
            url = 'media://' + url
        else:
            url = 'static://reports/images/placeholder.png'

        return url

then, in your templates:

.. code:: html

    <img class="pageLogoMiddle" src="{{'report-header-middle'|local_image_url}}">

where `'report-header-middle'` is the slug used to select the image.


Adding a plot to the PDF documents
==================================

In the frontend, you have many javascript libraries available to plot data and
draw fancy charts.

This doesn't help you in embedding a plot in a PDF documents built offline, however;
in this case, you need to build an image server side.

An helper function has been included in this app for that purpose; to use it, **matplotlib**
must be installed.

At the moment, it is more a POC then a complete solution; you can either use it
from the package, or copy the source file `pdf/plot.py` in your project and use
`build_plot_from_data()` as a starting point:

.. code:: python

    def build_plot_from_data(data, chart_type='line', as_base64=False, dpi=300, ylabel=''):
        """
        Build a plot from given "data";
        Returns: a bitmap of the plot

        Requires:
            matplotlib

        Keyword arguments:
        data -- see sample_line_plot_data() for an example; if None, uses sample_line_plot_data()
        chart_type -- 'line', 'bar', 'horizontalBar', 'pie', 'line', 'doughnut',
        as_base64 -- if True, returns the base64 encoding of the bitmap
        dpi -- bitmap resolution
        ylabel -- optional label for Y axis

        Data layout
        ===========

        Similar to django-jchart:

        - either (shared values for x)

            {
                "labels": ["A", "B", ...],
                "x" [x1, x2, ...],
                "columns": [
                    [ay1, ay2, ...],
                    [by1, by2, ...],
                ],
                "colors": [
                    "rgba(64, 113, 191, 0.2)",
                    "rgba(191, 64, 64, 0.0)",
                    "rgba(26, 179, 148, 0.0)"
                ]
            }

        - or

            {
                "labels": ["A", "B", ..., ],
                "columns": [
                    [
                        {"x": ax1, "y": ay1 },
                        {"x": ax2, "y": ay2 },
                        {"x": ax3, "y": ay3 },
                    ], [
                        {"x": bx1, "y": by1 },
                        {"x": bx2, "y": by2 },
                    ], ...
                ],
                "colors": ["transparent", "rgba(121, 0, 0, 0.2)", "rgba(101, 0, 200, 0.2)", ]
            }

        """

then, in the view, add the resulting bitmap to context:

.. code:: python

    def get_context_data(self, **kwargs):
        context = super().get_context_data(**kwargs)
        try:
            from .plot import build_plot_from_data
            plot_image = build_plot_from_data(data=None, chart_type='line', as_base64=True)
            context.update({
                'plot_image': plot_image,
            })
        except:
            pass
        return context

In the template, render it as an embedded image:

.. code:: html

    <style>
        .plot {
            border: 1px solid #ccc;
            width: 18cm;
            height: 6cm;
            margin: 1.0cm 0;
        }
    </style>

    {% if plot_image %}
        <img class="plot" src="data:image/png;base64,{{plot_image}}">
    {% endif %}

Try it
------

The management command `build_test_pdf` can be used with the "--plot_data" switch to test the resulting image:

.. code:: bash

    python manage.py build_test_pdf test.png -o -p '{"labels": ["sin", "cos"], "x": [0.0, 0.5, 1.0, 1.5, 2.0, 2.5, 3.0, 3.5, 4.0, 4.5], "columns": [[0.0, 9.09, -7.57, -2.79, 9.89, -5.44, -5.37, 9.91, -2.88, -7.51], [20.0, -13.07, -2.91, 16.88, -19.15, 8.16, 8.48, -19.25, 16.68, -2.56]]}' --plot_font Tahoma

|

.. image:: screenshots/plots.png


CHANGELOG.rst
.. :changelog:

History
=======

v0.1.0
------
* Publish on Pypi as "django-weasypdf"
* Support for 'line', 'bar', 'horizontalBar', 'pie', 'line', 'doughnut' plots

v0.0.5
------
* Support for plot added (only "line" chart type, at the moment; requires matplotlib)
* Customization examples added to readme

v0.0.4
------
* render_to_stream() renamed as render_as_pdf_to_stream()
* render_as_html_to_string() method added

v0.0.3
------
* render_to_stream() method added to produce a PDF from a background task

v0.0.2
------
* Some cleanup

v0.0.1
------
* Project start


pdf/plot.py
import io
import base64
import json
import math
from PIL import Image as PILImage
import datetime


def build_plot_from_data(data, chart_type='line', as_base64=False, dpi=300, ylabel=''):
    """
    Build a plot from given "data";
    Returns: a bitmap of the plot

    Requires:
        matplotlib

    Keyword arguments:
    data -- see sample_line_plot_data() for an example; if None, uses sample_line_plot_data()
    chart_type -- 'line', 'bar', 'horizontalBar', 'pie', 'line', 'doughnut',
    as_base64 -- if True, returns the base64 encoding of the bitmap
    dpi -- bitmap resolution
    ylabel -- optional label for Y axis

    Data layout
    ===========

    Similar to django-jchart:

    - either (shared values for x)

        {
            "labels": ["A", "B", ...],
            "x" [x1, x2, ...],
            "columns": [
                [ay1, ay2, ...],
                [by1, by2, ...],
            ],
            "colors": [
                "rgba(64, 113, 191, 0.2)",
                "rgba(191, 64, 64, 0.0)",
                "rgba(26, 179, 148, 0.0)"
            ]
        }

    - or

        {
            "labels": ["A", "B", ..., ],
            "columns": [
                [
                    {"x": ax1, "y": ay1 },
                    {"x": ax2, "y": ay2 },
                    {"x": ax3, "y": ay3 },
                ], [
                    {"x": bx1, "y": by1 },
                    {"x": bx2, "y": by2 },
                ], ...
            ],
            "colors": ["transparent", "rgba(121, 0, 0, 0.2)", "rgba(101, 0, 200, 0.2)", ]
        }

    """

    if data == None:
        data = sample_line_plot_data()

    if False:
        image_content = build_random_image()
    else:
        image_content = None
        try:
            with io.BytesIO() as buffer:

                colors = data['colors'] if 'colors' in data else default_plot_colors()
                if len(colors)>0 and type(colors[0]) == list:
                    colors = colors[0]

                if chart_type == 'line':
                    _build_line_plot_image(data, colors, buffer, dpi=dpi, ylabel=ylabel)
                elif chart_type in ['bar', 'horizontalBar', ]:
                    _build_bar_plot_image(data, colors, buffer, dpi=dpi, ylabel=ylabel, horizontal=(chart_type == 'horizontalBar'))
                elif chart_type == 'pie':
                    _build_pie_plot_image(data, colors, buffer, dpi=dpi, ylabel=ylabel)
                elif chart_type == 'doughnut':
                    _build_doughnut_plot_image(data, colors, buffer, dpi=dpi, ylabel=ylabel)
                else:
                    raise Exception('Unknown chart_type "%s"' % chart_type)

                #buffer.seek(0)
                pil_image = PILImage.open(buffer)
                with io.BytesIO() as output:
                    pil_image.save(output, format="PNG")
                    image_content = output.getvalue()
        except Exception as e:
            print('ERROR in build_plot_from_data(): ' + str(e))
            raise

    if as_base64:
        return base64.b64encode(image_content).decode()

    return image_content


def sample_line_plot_data():

    def real_value(value, decimals=2):
        text = '%.*f' % (decimals, value)
        return float(text)

    n = 100
    data = {
        'labels': ["sin", "cos"],
        'x': [0.5 * i for i in range(n)],
        'columns': [
            [real_value(10.0 * math.sin(i/2.0)) for i in range(n)],
            [real_value(20.0 * math.cos(i/4.0)) for i in range(n)],
        ]
    }
    return data


def build_random_image():
    # For test purposes only
    from random import randint
    rgb = (randint(0, 255), randint(0, 255), randint(0, 255))
    image_size = (200, 200)
    image = PILImage.new('RGB', image_size, rgb)
    with io.BytesIO() as output:
        image.save(output, format="PNG")
        image_content = output.getvalue()
    return image_content


def default_plot_colors(n = None):
    default_colors = [
        # Dygraphs defaults
        #"rgb(0,85,128)", "rgb(85,128,0)", "rgb(0,0,128)", "rgb(0,128,0)", "rgb(85,0,128)", "rgb(0,128,85)", "rgb(128,0,85)", "rgb(128,85,0)"
        #'#97742F', '#015480', '#648B18', '#292894', '#007F01', '#4C0E72', '#008055', '#800055',

        # CanvasJS defaults
        '#369EAD', '#C24642', '#7F6084', '#86B402', '#52514E', '#C8B631', '#6DBCEB', '#52514E',
    ]

    if n is None:
        colors = default_colors[:]
    else:
        colors = []
        m = len(default_colors)
        for i in range(n):
            colors.append(default_colors[i % m])
    return colors

def _format_x_time(seconds, x):
    text = str(datetime.timedelta(seconds=seconds))
    #remove leading "0:"
    if text.startswith('0:'):
        text = text[2:]
    return text


def _plot_color(plot_colors, index):

    def _rgb_string_to_color_code(rgb_string):
        # Accept either rgb(r,g,b) or #colorcode
        if rgb_string.startswith('#'):
            return rgb_string
        try:
            left_brace_position = rgb_string.find('(')
            right_brace_position = rgb_string.find(')')
            content = rgb_string[left_brace_position + 1:right_brace_position]
            tokens = [t.strip() for t in content.split(',')]
            color_code = '#%02x%02x%02x' % (int(tokens[0]), int(tokens[1]), int(tokens[2]), )
        except:
            color_code = '#000000'
        return color_code

    color_code = '#000000'
    if len(plot_colors) > 0:
        try:
            color = plot_colors[index % len(plot_colors)]
            color_code = _rgb_string_to_color_code(color)
        except:
            color_code = '#000000'
    return color_code


def _build_line_plot_image(plot_data, plot_colors, output_buffer, dpi, ylabel):
    """
    Read rectangular data from given "input_filename" CSV file, and build a bitmap with the plot;
    the result is saved in "output_filename"
    """

    # Solve "'NSWindow drag regions should only be invalidated on the Main Thread!' - macos/python"
    # https://github.com/matplotlib/matplotlib/issues/14304#issuecomment-545717061
    import matplotlib
    from matplotlib import pyplot as plt

    matplotlib.use('agg')

    if 'x' in plot_data:
        x_shared = plot_data['x']
    else:
        x_shared = None

    formatter = matplotlib.ticker.FuncFormatter(_format_x_time)
    fig, axes = plt.subplots(nrows=1, ncols=1, figsize=(8,3), dpi=dpi, tight_layout=True)
    axes.xaxis.set_major_formatter(formatter)
    for index, values in enumerate(plot_data['columns']):

        if x_shared is not None:
            x = x_shared
            y = values
        else:
            x = [item['x'] for item in values]
            y = [item['y'] for item in values]

        #axes.plot(x, y, color, linewidth=1)
        assert len(x) == len(y)
        color = _plot_color(plot_colors, index)
        axes.plot(x, y, color, linewidth=1)
        # # add this column to chart

    plt.ylabel(ylabel)
    if plot_data.get('labels', []):
        plt.legend(plot_data['labels'])
    plt.grid(axis='y', color='#ccc')

    #plt.grid()
    plt.savefig(output_buffer)  # , dpi=600)


def ellipsis(text, max_length):
    if text is None:
        return ''
    text = str(text)
    if len(text) > max_length:
        text = text[:max_length - 1] + '…'
    return text


def _build_bar_plot_image(plot_data, plot_colors, output_buffer, dpi, ylabel, horizontal=False):
    import matplotlib
    from matplotlib import pyplot as plt

    # Solve "'NSWindow drag regions should only be invalidated on the Main Thread!' - macos/python"
    # https://github.com/matplotlib/matplotlib/issues/14304#issuecomment-545717061
    matplotlib.use('agg')

    fig, axes = plt.subplots(nrows=1, ncols=1, figsize=(8, 6), dpi=dpi, tight_layout=True)

    values = plot_data['columns'][0]
    labels = [ellipsis(l, 20) for l in plot_data['x']]
    assert len(values) == len(labels)

    colors = [_plot_color(plot_colors, index) for index in range(len(values))]
    assert len(values) == len(colors)

    # Higher values at top
    values.reverse()
    labels.reverse()
    colors.reverse()

    if horizontal:
        axes.barh(labels, values, color=colors)
    else:
        axes.bar(labels, values, color=colors)

    plt.xticks(fontsize=10)
    plt.yticks(fontsize=10)

    #
    # https://mode.com/example-gallery/python_horizontal_bar/
    #

    # Despine
    axes.spines['right'].set_visible(False)
    axes.spines['top'].set_visible(False)
    axes.spines['left'].set_visible(False)
    axes.spines['bottom'].set_visible(False)

    # Draw vertical axis lines
    vals = axes.get_xticks()
    for tick in vals:
        axes.axvline(x=tick, linestyle='solid', alpha=0.4, color='#eeeeee', zorder=1)

    # # Switch off ticks
    # axes.tick_params(axis="both", which="both", bottom="off", top="off", labelbottom="on", left="off", right="off", labelleft="on")

    # # Format y-axis label
    # axes.xaxis.set_major_formatter(matplotlib.ticker.StrMethodFormatter('{x:,g}'))


    #plt.legend(plot_data['labels'], fontsize=10, )
    #plt.ylabel(ylabel)
    #plt.grid(axis='y', color='#ccc')
    plt.savefig(output_buffer)  # , dpi=600)


def pct_func(pct, allvals):
    #absolute = int(pct/100.*np.sum(allvals))
    absolute = int(pct/100. * sum(allvals))
    #return "{:.1f}%\n({:d} g)".format(pct, absolute)
    return "{:d}\n({:.1f}%)".format(absolute, pct)


def _build_pie_plot_image(plot_data, plot_colors, output_buffer, dpi, ylabel):

    import matplotlib
    from matplotlib import pyplot as plt

    # Solve "'NSWindow drag regions should only be invalidated on the Main Thread!' - macos/python"
    # https://github.com/matplotlib/matplotlib/issues/14304#issuecomment-545717061
    matplotlib.use('agg')

    # Adapted from:
    # https://matplotlib.org/3.1.1/gallery/pie_and_polar_charts/pie_and_donut_labels.html
    fig, axes = plt.subplots(nrows=1, ncols=1, figsize=(8, 6), dpi=dpi, tight_layout=True, subplot_kw=dict(aspect="equal"))

    values = plot_data['columns'][0]
    labels = plot_data['x']
    colors = [_plot_color(plot_colors, index) for index in range(len(values))]
    assert len(values) == len(labels)
    assert len(values) == len(colors)
    # values.reverse()
    # labels.reverse()
    # colors.reverse()

    wedges, texts, autotexts = axes.pie(values, autopct=lambda pct: pct_func(pct, values), textprops=dict(color="w"), colors=colors, startangle=90)
    axes.legend(wedges, labels, title="", loc="center left", bbox_to_anchor=(1, 0, 0.5, 1))
    plt.setp(autotexts, size=8, weight="bold")

    if ylabel:
        axes.set_title(ylabel)

    plt.savefig(output_buffer)  # , dpi=600)


def _build_doughnut_plot_image (plot_data, plot_colors, output_buffer, dpi, ylabel):

    import matplotlib
    from matplotlib import pyplot as plt
    import numpy as np

    # Solve "'NSWindow drag regions should only be invalidated on the Main Thread!' - macos/python"
    # https://github.com/matplotlib/matplotlib/issues/14304#issuecomment-545717061
    matplotlib.use('agg')

    # Adapted from:
    # https://matplotlib.org/3.1.1/gallery/pie_and_polar_charts/pie_and_donut_labels.html
    fig, axes = plt.subplots(nrows=1, ncols=1, figsize=(8, 6), dpi=dpi, tight_layout=True, subplot_kw=dict(aspect="equal"))

    values = plot_data['columns'][0]
    labels = plot_data['x']
    colors = [_plot_color(plot_colors, index) for index in range(len(values))]
    assert len(values) == len(labels)
    assert len(values) == len(colors)
    wedges, texts, autotexts = axes.pie(
        values,
        autopct=lambda pct: pct_func(pct, values),
        pctdistance=0.75,
        #labeldistance=1.2,
        wedgeprops=dict(width=0.5),
        startangle=-40,
        colors=colors
    )

    bbox_props = dict(boxstyle="square,pad=0.3", fc="w", ec="k", lw=0.72)
    kw = dict(arrowprops=dict(arrowstyle="-"), bbox=bbox_props, zorder=0, va="center")

    for i, p in enumerate(wedges):
        ang = (p.theta2 - p.theta1)/2. + p.theta1
        y = np.sin(np.deg2rad(ang))
        x = np.cos(np.deg2rad(ang))
        horizontalalignment = {-1: "right", 1: "left"}[int(np.sign(x))]
        connectionstyle = "angle,angleA=0,angleB={}".format(ang)
        kw["arrowprops"].update({"connectionstyle": connectionstyle})
        #axes.annotate(recipe[i], xy=(x, y), xytext=(1.35*np.sign(x), 1.4*y), horizontalalignment=horizontalalignment, **kw)
        axes.annotate(labels[i], xy=(x, y), xytext=(1.35*np.sign(x), 1.4*y), horizontalalignment=horizontalalignment, **kw)

    if ylabel:
        axes.set_title(ylabel)
    plt.savefig(output_buffer)  # , dpi=600)


pdf/management/commands/build_test_pdf.py
# -*- coding: UTF-8 -*-
from __future__ import print_function
import os
import datetime
import json
import pprint
import argparse
from django.core.management.base import BaseCommand
from django.conf import settings
from pdf.utils import build_pdf_document
from pdf.views import PdfTestView
from pdf.plot import build_plot_from_data

        # fonts = sorted(set([f.name for f in matplotlib.font_manager.fontManager.ttflist]))
        # print(fonts)

        # rcParams['font.family'] = 'Tahoma'

class Command(BaseCommand):
    help = "Build a test PDF document (or plot)"
    epilog = """
Sample usages:


python manage.py build_test_pdf test.pdf -o


python manage.py build_test_pdf test.png -o -p '{"labels": ["sin", "cos"], "x": [0.0, 0.5, 1.0, 1.5, 2.0, 2.5, 3.0, 3.5, 4.0, 4.5], "columns": [[0.0, 9.09, -7.57, -2.79, 9.89, -5.44, -5.37, 9.91, -2.88, -7.51], [20.0, -13.07, -2.91, 16.88, -19.15, 8.16, 8.48, -19.25, 16.68, -2.56]]}' --plot_font Tahoma


python manage.py build_test_pdf test.png -o --plot_type horizontalBar -p '{"labels": [null], "columns": [[1085, 839, 340, 285, 270, 211, 210, 177, 161, 141, 139, 137, 137, 134, 124, 122, 119, 116, 112, 108]], "colors": [["#f2f6ee", "#353a3a", "#fcfcfc", "#a23c3a", "#ffd100", "#0f6fa8", "#00968f", "#33b3ad", "#c80000", "#a83e3b", "#244f28", "#b8c0c2", "#586062", "#655580", "#6fd3d1", "#d0cd1d", "#cacac5", "#828b88", "#148396", "#bec8d1"]], "x": ["CW051W", "AC144N", "0011/Sugar Dust-EA Semibrillo Sipa Alfa-Sipamundo-Base Master", "AC118R", "AC105Y", "AC079N", "7116A", "7115D", "1109/Party Time-EA Semibrillo Sipa Alfa-Sipamundo-Base Master", "AC111R", "0535/Zen Retreat-EA Semibrillo SipaAlfa-Sioamundo-BaseMaster", "0546/Paternoster-EA Semibrillo Sipa Alfa-Sipamundo-Base Master", "0534/Subtle Shadow-EA Semibrillo Sipa Alfa-Sipamundo-Base Master", "7006N", "7114M", "7216N", "8792W/Monorail Silver CGl -E.A. P. y F. BioTech-Fandex Millennium-Base Blanca", "0533/Techile-EA Semibrillo Sipa Alfa-Sipamundo-Base Master", "0682/Pleasant Stream-EA Semibrillo Sipa Alfa-Sipamundo-Base Master", "0510/Sacred Spring-EA Semibrillo Sipa Alfa-Sipamundo-Base Master"]}'


python manage.py build_test_pdf test.png -o -p '{"labels": ["Total", "Custom", "Stock"], "columns": [[4, 13, 101, 78, 94, 152, 198, 337, 329, 290, 278, 295, 448, 353, 326, 512, 559, 408, 512, 440, 815, 1390, 819, 1562, 2615, 2217, 2352, 1929, 3469, 5791, 6900, 6195], [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 25, 0, 0, 14, 36, 67, 8, 46, 67, 519, 585, 559, 433], [4, 13, 101, 78, 94, 152, 198, 337, 329, 290, 278, 295, 448, 353, 326, 512, 559, 408, 512, 415, 815, 1390, 805, 1526, 2548, 2209, 2306, 1862, 2950, 5206, 6341, 5762]], "colors": ["rgba(64, 113, 191, 0.2)", "rgba(191, 64, 64, 0.0)", "rgba(26, 179, 148, 0.0)"], "x": ["03/01/2018", "04/01/2018", "05/01/2018", "06/01/2018", "07/01/2018", "08/01/2018", "09/01/2018", "10/01/2018", "11/01/2018", "12/01/2018", "01/01/2019", "02/01/2019", "03/01/2019", "04/01/2019", "05/01/2019", "06/01/2019", "07/01/2019", "08/01/2019", "09/01/2019", "10/01/2019", "11/01/2019", "12/01/2019", "01/01/2020", "02/01/2020", "03/01/2020", "04/01/2020", "05/01/2020", "06/01/2020", "07/01/2020", "08/01/2020", "09/01/2020", "10/01/2020"]}'


python manage.py build_test_pdf test.png -o --plot_type pie -p '{"labels": [], "columns": [[1498, 1217, 528, 479, 363, 353]], "colors": [["rgb(255, 99, 132)", "rgb(255, 159, 64)", "rgb(255, 205, 86)", "rgb(75, 192, 192)", "rgb(54, 162, 235)", "rgb(153, 102, 255)", "rgb(201, 203, 207)"]], "x": ["255: TOO LOW WATER LEVEL", "210: DOOR OPEN", "211: COVERS NOT AVAILABLE", "2000: Paper end", "1000: Scale not responding", "895: CAN_ABSENT_DURING_DISPENSING"]}'


python manage.py build_test_pdf test.png -o --plot_type doughnut -p '{"labels": [], "columns": [[1498, 1217, 528, 479, 363, 353]], "colors": [["rgb(255, 99, 132)", "rgb(255, 159, 64)", "rgb(255, 205, 86)", "rgb(75, 192, 192)", "rgb(54, 162, 235)", "rgb(153, 102, 255)", "rgb(201, 203, 207)"]], "x": ["255: TOO LOW WATER LEVEL", "210: DOOR OPEN", "211: COVERS NOT AVAILABLE", "2000: Paper end", "1000: Scale not responding", "895: CAN_ABSENT_DURING_DISPENSING"]}'

"""

    def create_parser(self, prog_name, subcommand, **kwargs):
        if self.epilog:
            kwargs.update({
                'epilog': self.epilog,
            })
        parser = super().create_parser(prog_name, subcommand, **kwargs)
        if self.epilog:
            parser.formatter_class =  argparse.RawTextHelpFormatter
        return parser

    def add_arguments(self, parser):
        parser.add_argument("filepath", help='ouput file path (use .pdf or .png extension as required')
        parser.add_argument('--open', '-o', action='store_true', help="open resulting file")
        parser.add_argument("--plot_data", "-p", help='JSON plot data to build a bitmap (instead of a pdf)')
        parser.add_argument('--plot_type', "-t", choices=['line', 'bar', 'horizontalBar', 'pie', 'doughnut', ], default="line")
        parser.add_argument('--plot_font', )
        parser.add_argument('--list_fonts', "-l", action='store_true', help="list matplotlib fonts and exit")

    def list_fonts(self):
        import matplotlib
        fonts = list(set([f.name for f in matplotlib.font_manager.fontManager.ttflist]))
        sorted_fonts = sorted(fonts, key=str.casefold)
        pprint.pprint(sorted_fonts)

    def select_font(self, font):
        import matplotlib
        matplotlib.rcParams['font.family'] = font

    def handle(self, *args, **kwargs):

        if kwargs['list_fonts']:
            # list fonts and exit
            self.list_fonts()
            return

        if kwargs['plot_font']:
            self.select_font(kwargs['plot_font'])

        # Use either user supplied filepath, or the a default name provided by the view
        expected_file_extension = '.pdf'
        filepath = kwargs['filepath']
        filename, file_extension = os.path.splitext(filepath)

        # Retrieve and deserialized user-supplied plot data
        plot_data = kwargs['plot_data']
        if plot_data is not None:
            plot_data = json.loads(plot_data)
            expected_file_extension = '.png'

        # File cleanup
        assert file_extension.lower() == expected_file_extension, "Wrong extension: %s" % file_extension
        if os.path.isfile(filepath):
            os.remove(filepath)

        with open(filepath, 'wb') as f:
            if plot_data is not None:
                image = build_plot_from_data(
                    plot_data,
                    chart_type=kwargs['plot_type'],
                    as_base64=False,
                    dpi=300,
                    ylabel='test plot',
                )
                f.write(image)
            else:
                # Create a View to work with
                view = PdfTestView()
                context = view.get_context_data()
                view.render_as_pdf_to_stream('', context, f)

        print('Result in: "%s"' % os.path.abspath(filepath))
        if kwargs['open']:
            os.system('open %s' % filepath)
